### ISSUE TITLE
How to unwrap after auto_wrap in FSDP?

### ISSUE THREAD

**Issue by ZN1010** [pytorch/pytorch#103962]
I am currently fine-tuning a LLM (LLaMA) and would like to retrieve the gradients of each weight (parameter) after every gradient update. However, I notice that weights are (auto) wrapped into stuff like “_fsdp_wrapped_module._flat_param” during training. I need to map these wrapped weights to the original LLaMA architecture such as “self_attn.v_proj”. Any code examples?

I guess “summon_full_params()” might be the function that I look for, but I am not sure if that is correct. I also have difficulty using this function. Thanks a lot for any help!

**Comment by awgu** [comment]
Unfortunately, we do not have great support for gathering the gradients. I added an option like `summon_full_params(model, offload_to_cpu=False, with_grads=True)` (note the `offload_to_cpu=False`), which requires passing `use_orig_params=True` to the FSDP constructor.

If you have significant memory pressure, then the `offload_to_cpu=False` may be problematic. In that case, you can also try `summon_full_params(..., recurse=False)` to gather the gradients layer by layer.

**Comment by rahul-1996** [comment]
Did you find a good way to accomplish this? Stuck in a similar situation.

**Comment by awgu** [comment]
> Did you find a good way to accomplish this? Stuck in a similar situation.

We are working on some new things. It would be great if you could tell me more about your use case. Thanks!
